### PR TITLE
Fix editor live validation

### LINKED ISSUE #5533: Editor Code Check not 'Live' anymore [Edit: Being worked on]
### The problem

When editing code, the editor no longer updates live with errors. 
 - You have to save, close and re-open for them to appear.
 

### Which version of ESPHome has the issue?

2024.2.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP8266

### Board

_No response_

### Component causing the issue

_No response_

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This broke in 25ab6f0297fcf3a21e07644854db5083ba5a0eed because `read_config_file` was no longer being called. It was missed that `read_config_file` will check the `CORE.vscode` and `CORE.ace` globals, and in some cases instead of reading a config file from disk, it will print an event to stdout, and than decode json from stdin to get the contents of the yaml.

```python
def read_config_file(path: str) -> str:
    if CORE.vscode and (
        not CORE.ace or os.path.abspath(path) == os.path.abspath(CORE.config_path)
    ):
```

This PR removes `read_config_file` since its no longer used, which removes more of the global `CORE.ace` and `CORE.vscode` checks.

There are still two places where these global checks remain that should be removed, but they are a bit harder to remove and not contemplated in this PR
```python
esphome/config_validation.py:        not CORE.ace or os.path.abspath(path) == os.path.abspath(CORE.config_path)
esphome/config_validation.py:        not CORE.ace or os.path.abspath(path) == os.path.abspath(CORE.config_path)
```

I added some more typing since it was very hard to follow what was going on as the variables names were not obvious.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/5533

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/vscode.py [hunk 1]
```diff
@@ -56,6 +58,25 @@ def add_validation_error(self, range_, message):
...
+        json.dumps(
+            {
+                "type": "read_file",
+                "path": path,
+            }
+        )
+    )
+
+
 def read_config(args):
```
Comment (bdraco) [inline esphome/vscode.py:80]: It would be nice to add tests for this, but in its current form its not very testable, and I didn't want to refactor it here.

**Review by jesserockz — APPROVED** [review]
